### PR TITLE
feat: docker exclude base image vulns flag

### PR DESCRIPTION
**Pull request by karniwl**
- [x] Ready for review
- [x] Follows [CONTRIBUTING](https://github.com/snyk/snyk/blob/master/.github/CONTRIBUTING.md) rules
- [x] Reviewed by Snyk internal team

#### What does this PR do?
- Introduces `user layer` flag, when used, only user layer issues will be displayed.
- fixes single/multiple vulnerabilities text for key binaries display.

#### Screenshot
image

added `Pro tip` to introduce flag:
image

### CODE REVIEW

File: src/cli/commands/test.ts [hunk 1]
```diff
@@ -249,16 +249,42 @@ function displayResult(res, options) {
   }
   let summary = testedInfoText + ', ' + chalk.red.bold(vulnCountText);
 
+  if (options['user-layer'] && res.vulnerabilities) {
```
Comment (orkamara) [inline src/cli/commands/test.ts]: do we need to check if this is `docker` case?

File: src/cli/commands/test.ts [hunk 2]
```diff
@@ -249,16 +249,42 @@ function displayResult(res, options) {
...
+    userUniqueCount = res.vulnerabilities.reduce((acc, curr) => {
+      if (!seen[curr.id]) {
+        seen[curr.id] = true;
+        acc++;
+      }
+      return acc;
+    }, 0);
+    const userLayerVulnText = userUniqueCount === 1 ? 'vulnerability' : 'vulnerabilities';
+    const userLayerText = '\nOut of which ' + chalk.bold.red(`${userUniqueCount} ${userLayerVulnText}`) +
+    ' introduced in the Docker user layer.';
```
Comment (orkamara) [inline src/cli/commands/test.ts]:
```suggestion
    ' introduced in the Docker user layers.';
```

File: src/cli/commands/test.ts [hunk 3]
```diff
@@ -249,16 +249,42 @@ function displayResult(res, options) {
...
 
   if (options.docker &&
-    !options.file &&
     (config.disableSuggestions !== 'true')) {
-    dockerSuggestion += chalk.bold.white('\n\nPro tip: use `--file` option to get base image remediation advice.' +
-      `\nExample: $ snyk test --docker ${options.path} --file=path/to/Dockerfile` +
-      '\n\nTo remove this message in the future, please run `snyk config set disableSuggestions=true`');
+    const optOutSuggestions =
+    '\n\nTo remove this message in the future, please run `snyk config set disableSuggestions=true`';
+    if (!options.file) {
```
Comment (orkamara) [inline src/cli/commands/test.ts]: maybe we don't want to display pro tip for both `--file` and the new flag in the same time?
So maybe add an `else`

File: src/cli/commands/test.ts [hunk 4]
```diff
@@ -249,16 +249,42 @@ function displayResult(res, options) {
...
+      dockerSuggestion += chalk.bold.white('\n\nPro tip: use `--file` option to get base image remediation advice.' +
+      `\nExample: $ snyk test --docker ${options.path} --file=path/to/Dockerfile`);
+    }
+    if (!options['user-layer']) {
```
Comment (orkamara) [inline src/cli/commands/test.ts]: Nice!!

File: src/cli/commands/test.ts [hunk 5]
```diff
@@ -249,16 +249,42 @@ function displayResult(res, options) {
...
+      dockerSuggestion +=
+      chalk.bold.white('\n\nNew flag: use `--user-layer` to display user layer vulnerabilites only.');
```
Comment (orkamara) [inline src/cli/commands/test.ts]:
```suggestion
      chalk.bold.white('\n\nPro tip: use `--user-layer` to display user layers vulnerabilites only.');
```

File: help/help.txt [hunk 6]
```diff
@@ -43,6 +43,8 @@ Options:
                        vulnerabilities. Can be used alongside `--file` and a
                        path to the image's Dockerfile for more detailed
                        remediation advice.
+  --file ............. Specify path to Dockerfile to get base image remediation advice.
```
Comment (orkamara) [inline help/help.txt]: there is already an `--file` documentation (Can be used alongside `--file` and a path to the image's Dockerfile for more detailed remediation advice).

File: help/help.txt [hunk 7]
```diff
@@ -43,6 +43,8 @@ Options:
...
+  --user-layer ....... Display vulnerabilities introduced by Docker user layer only.
```
Comment (orkamara) [inline help/help.txt]:
```suggestion
  --user-layers ....... Display vulnerabilities introduced by Docker user layers only.
```

File: src/cli/commands/test.ts [hunk 8]
```diff
@@ -272,8 +298,9 @@ function displayResult(res, options) {
     .filter((vuln) => (vuln.metadata.packageManager !== 'upstream'));
   const binariesSortedGroupedVulns = sortedGroupedVulns
     .filter((vuln) => (vuln.metadata.packageManager === 'upstream'));
+
   const groupedVulnInfoOutput = filteredSortedGroupedVulns.map((vuln) => formatIssues(vuln, options));
-  const groupedDockerBinariesVulnInfoOutput = (res.docker && res.docker.binariesVulns) ?
+  const groupedDockerBinariesVulnInfoOutput = (res.docker && binariesSortedGroupedVulns.length) ?
```
Comment (orkamara) [inline src/cli/commands/test.ts]: what's the reason for this change?

File: src/cli/commands/test.ts [hunk 9]
```diff
@@ -249,16 +249,42 @@ function displayResult(res, options) {
   }
   let summary = testedInfoText + ', ' + chalk.red.bold(vulnCountText);
 
+  if (options['user-layer'] && res.vulnerabilities) {
+    res.vulnerabilities =  res.vulnerabilities.filter((vuln) => (vuln.dockerfileInstruction));
+    let userUniqueCount = 0;
+    const seen = {};
+    userUniqueCount = res.vulnerabilities.reduce((acc, curr) => {
```
Comment (orkamara) [inline src/cli/commands/test.ts]: Can you explain why do we need unique count?

File: src/cli/commands/test.ts (hunk 8 above)
Comment (karniwl) [reply src/cli/commands/test.ts]: to display `1 vulnerability` and not `1 vulnerabilities` in case of 1 vulnerabilty

File: help/help.txt (hunk 6 above)
Comment (karniwl) [reply help/help.txt]: I don't understand your suggestion...

File: src/cli/commands/test.ts (hunk 8 above)
Comment (karniwl) [reply src/cli/commands/test.ts]: see in pr description

File: src/cli/commands/test.ts [hunk 10]
```diff
@@ -249,16 +249,27 @@ function displayResult(res, options) {
...
+      '\n\nTo remove this message in the future, please run `snyk config set disableSuggestions=true`';
+    if (!options.file) {
+      dockerSuggestion += chalk.bold.white('\n\nPro tip: use `--file` option to get base image remediation advice.' +
+        `\nExample: $ snyk test --docker ${options.path} --file=path/to/Dockerfile`);
+    } else if (!options['exclude-base-image-vulns']) {
+      dockerSuggestion +=
+        chalk.bold.white(
+          '\n\nPro tip: use `--exclude-base-image-vulns` to exclude from display Docker base image vulnerabilities.');
+    }
+    if (!options.file || !options['exclude-base-image-vulns']) {
```
Comment (orkamara) [inline src/cli/commands/test.ts]: move to the if part above

File: src/cli/commands/test.ts [hunk 11]
```diff
@@ -625,3 +638,26 @@ function metadataForVuln(vuln) {
     packageManager: vuln.packageManager,
   };
 }
+
+function getDockerLayersVulnCount(options, res): string {
+  if (options.docker && options.file && res.vulnerabilities) {
```
Comment (orkamara) [inline src/cli/commands/test.ts]: flip the if :) (as we talked about)
